docs/RFCS/20170925_jsonb_scope.md

```diff
@@ -0,0 +1,123 @@
+for CockroachDB, with the
+[issue](https://github.com/cockroachdb/cockroach/issues/2969) having by far
+the largest number of thumbs-ups of any feature request.
+Many users of CockroachDB would prefer to store JSON blobs in columns, either
+in addition to a traditional relational schema, or as a replacement of it
+(storing all data in a single JSON column, plus a primary key).
+
+Reasons one might want to use JSON with CockroachDB include:
+* Ease of storing and manipulating existing JSON data, without having to do a
+  manual conversion to a relational schema
+* Storing data that the schema author has little-to-no control over in a
+  structured, queryable way
+* Having access to CockroachDB's transactions and strong consistency while
+  working with the ergonomics of a Mongo-style loose-schema database
+
+## Design
+
+Postgres supports two data types - `JSON` and `JSONB`.
+Our datatype will be more similar to `JSONB` in its behaviour, and as a result we
+will refer to it as `JSONB`, rather than `JSON`.
+`JSON` in Postgres refers to JSON data that is stored as text,
+preserving whitespace, duplicate keys, and the ordering of keys, and only
+verifying that the value of that column is valid JSON. `JSONB` stores only
+the actual meaning of the JSON value, throwing away whitespace, duplicate
+keys, and key ordering, allowing for more efficient storage, and removing the
+need to parse the entire object to retrieve a specific field.
+
+There are two main directions we could go with the implementation of JSONB -
+broadly, we could split JSON objects across kv entries or put them entirely in a
+single kv entry.
+This is a material decision as it affects what can plausibly be accomplished
+with JSONB columns.
+We opt to store JSONB blobs entirely within a single kv entry, with an eye
+towards keeping the door open to split them up if this becomes a bottleneck for
+users.
+
+## Details
+
+### Operators
+
+We plan to support all of Postgres’s JSON manipulation and access operators
+(`->`, `->>`, `||`, `-`, `#-`), in addition to the ones which can be accelerated by
+inverted indexes (`@>`, `?`, `?|`, `?&`).
+
+The behaviour of each of these operators is described below:
+
+| Operator     | Description                                           | Example                                                                    |
+| ------------ | ----------------------------------------------------- | -------------------------------------------------------------------------- |
+| `->`         | Access a JSON field, returning a JSON value.          | `'{"foo":"bar"}'::JSONB->'foo'` = `'bar'::JSONB`                           |
+| `->>`        | Access a JSON field, returning a string.              | `'{"foo":"bar"}'::JSONB->>'foo'` = `'bar'::STRING`                         |
+| `||`         | Concatenate JSON arrays, or append to a JSON array.   | `'[1, 2]'::JSONB || '[3]'::JSONB` = `'[1, 2, 3]'::JSONB`                   |
```

Comment:
This row is broken in the rendered output because `|` is interpreted as a Markdown table column separator. If you care, I think you can fix it by using `<code>&#124;&#124;</code>` instead at the cost of making the plain-text version unreadable.